File: Multiton.java
import java.util.HashMap;
import java.util.Map;

public class Multiton {
    // Armazena instâncias únicas por chave
    private static final Map<String, Multiton> instances = new HashMap<>();

    private String name;

    // Construtor privado
    private Multiton(String name) {
        this.name = name;
        System.out.println("Instância Multiton criada: " + name);
    }

    // Método público para obter a instância por chave
    public static synchronized Multiton getInstance(String key) {
        if (!instances.containsKey(key)) {
            instances.put(key, new Multiton(key));
        }
        return instances.get(key);
    }

    public void showName() {
        System.out.println("Instância: " + name);
    }

    public static void main(String[] args) {
        Multiton m1 = Multiton.getInstance("DB1");
        Multiton m2 = Multiton.getInstance("DB2");
        Multiton m3 = Multiton.getInstance("DB1"); // Reutiliza a instância de "DB1"

        m1.showName(); // Instância: DB1
        m2.showName(); // Instância: DB2

        // Verifica se as instâncias são iguais para a mesma chave
        System.out.println(m1 == m3); // true
    }

}

File: Multiton.js
/*
Em um sistema grande, diferentes módulos (exemplo: banco de dados, interface gráfica, autenticação) 
podem ter suas próprias configurações. O Multiton permite que cada módulo tenha uma única instância 
de configuração, evitando recriação desnecessária e garantindo que tenha uma consistência bacana.
*/

class ConfigManager {
    static instances = {}; // Armazena instâncias únicas para cada módulo

    constructor(moduleName) {
        if (ConfigManager.instances[moduleName]) {
            return ConfigManager.instances[moduleName];
        }

        this.moduleName = moduleName;
        this.settings = {}; // Armazena as configurações do módulo

        ConfigManager.instances[moduleName] = this;
    }

    setConfig(key, value) {
        this.settings[key] = value;
    }

    getConfig(key) {
        return this.settings[key] || null;
    }
}

// Criando configurações para diferentes módulos
const dbConfig = new ConfigManager("database");
dbConfig.setConfig("host", "localhost");
dbConfig.setConfig("port", 5432);

const uiConfig = new ConfigManager("ui");
uiConfig.setConfig("theme", "dark");
uiConfig.setConfig("language", "en");

const authConfig = new ConfigManager("auth");
authConfig.setConfig("tokenExpiration", "1h");

// Recuperando a mesma instância para "database"
const dbConfigAgain = new ConfigManager("database");
console.log(dbConfigAgain.getConfig("host")); // "localhost"
console.log(dbConfigAgain === dbConfig); // true (mesma instância)

// Recuperando a mesma instância para "ui"
const uiConfigAgain = new ConfigManager("ui");
console.log(uiConfigAgain.getConfig("theme")); // "dark"

console.log(ConfigManager.instances); // Exibe todas as instâncias gerenciadas pelo Multiton
